### ISSUE TITLE
Notes from Dynamicland: Geokit

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10601]
Notes from Dynamicland: Geokit<br>
<br>
https://ift.tt/hHVTmlE<br>
<br>
<br>
<br>

<div><div><div><div>
<div>
<p>Hey ___,</p>
<p>I know you've been curious about some of the stuff I've done at
<a href="https://dynamicland.org/">Dynamicland</a>. It might be a while until you
can make it out here and visit in person, so I'll try to explain one
project of mine, Geokit, from top to bottom, to give you some of the
flavor.</p>
<p>I'll discuss what Geokit actually is and what you do with it
first, and then I'll explain how it's made. That explanation might
help you understand how people make things in Dynamicland in
general. It's the kind of detail that isn't covered in the zine or on
the Dynamicland website.</p>
<p>I sort of hope you read this, come here yourself, and reimplement
Geokit better.</p>
<h2>What Geokit is</h2>
<p>Geokit is a 'kit' or 'library' for building and viewing maps.</p>
<p>One Dynamicland researcher, <a href="https://twitter.com/lukexi">Luke
Iannini</a>, writes about his experiences
with Geokit:</p>
<blockquote>
<p>The real estate group that owns the building Dynamicland lives in
uses their purchasing power to influence the city to improve
infrastructure to the underserved areas of Oakland they develop
in. The head of this initiative was visiting the space, and she
began describing her work while we were gathered around the lunch
table, which happened to have Geokit spread across it looking at
Oakland. Without breaking the flow of conversation, I dealt the
transit card to display the bus routes, and she grabbed the “zoom
and pan dial”<sup></sup> without any instruction to zoom in on a portion
of West Oakland — noticing a huge hole in route coverage she’d never
seen before where she knew there were tons of working families. We
spent the next 15 minutes exploring as she taught me more about the
city's transit details than I ever knew I wanted to know.</p>
<p>In another case at a party I learned that a person I’d just met grew
up in the same city as me (Tucson, AZ) when she searched for it to
print it out, and we then bonded deeply over looking at the
incredibly stark racial divisions in North and South Tucson that we
knew from experience but had never seen laid out in raw data.</p>
<p>I’ve never had experiences like these in any other medium. Buildings
don’t tend to have detailed racial, transit and elevation maps of
every city in the US on the walls. In theory you could pull this
data up on your personal iPad, praying there aren’t any embarrassing
notifications and hopefully remembering what you were even going to
look at by the time you’re past the lock screen looking at 24
rainbow-gradient icons and red bubbles, but I am personally not in
the habit of pulling out my devices in the middle of parties. In
both of these cases these just-dynamic-enough maps were <em>places</em> at
the party, just like the appetizers table and the piano, where
people could casually gather, play and converse.</p>
</blockquote>
<p>Geokit lives on a shelf when it's not in use, so I'll take it off the
shelf and lay it out on a table:</p>
<div>
<p>Taking Geokit off the shelf. Some other kits below it; <a href="https://twitter.com/smd4/status/1002410711354523648">general tools</a> on
the right</p>
</div>
<h3>Geokit in action</h3>
<p>I'm going to pull up a map of San Francisco and zoom into Dolores
Park.</p>
<p>I'll start by typing into the "Geomap above from place search" page,
which turns the giant blank page above it into a map.</p>
<div>
<video></video><p>Typing places into Geokit and seeing maps</p>
</div>
<p>If you were here in person, I'd hand <em>you</em> the keyboard, and I bet
you'd want to type your hometown in. But I'll just use San Francisco.</p>
<p>OK, now I want to zoom in to the park. I have a tool for zooming, the
"Map magnifier with IR (infrared) dial" page (the smaller page on the
right). When I put that page on the table, it becomes a zoomed-in
'inset map.'</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/ir-dial-labeled.jpg"><p>Inset map (right) shows a zoomed-in view of the dial's
location on the big map (left)</p>
</div>
<p>I also have to put this rotating dial on the big map of SF, which
controls both the location and zoom level of the inset map. Moving the
dial pans the inset map, and rotating the dial zooms it in and
out. Being able to
<a href="https://en.wikipedia.org/wiki/Direct_manipulation_interface">directly</a>
move the map position is really satisfying.</p>
<p>In this next clip, I put the dial down and then pan over and zoom the
inset map to Dolores Park in the Mission.</p>
<div>
<video></video><p>Using the dial to pan and zoom the inset map to Dolores Park</p>
</div>
<p>Now I'm curious about this part of the neighborhood. How diverse is
it? How has it changed over time? Are there stark dividing lines?
I'll apply some data layers: a color-coded demographic map (<a href="https://demographics.virginia.edu/DotMap/">"racial
dot map"</a>), and then street
lines and labels on top so the map is more understandable. I point
them at the big map first, then point them at the zoomed map (just to
show that both work).</p>
<div>
<video></video><p>Applying tilelayer overlays to the big map and then
just to the inset map</p>
</div>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/dolores-park-racial-dot-map.jpg"><p>Seeing demographics around Dolores Park by pointing
the "Demographic dots tilelayer" page at the inset map. The color legend is
written on that tilelayer page – an advantage of using real paper
rather than a screen</p>
</div>
<p>Check out how the west (toward Castro) is bluer on the inset map,
while the east (toward Mission) is more orange.<sup></sup> Imagine
having multiple people at the table: each person might have their
personal view on smaller pages, but the table can all share the big
map at the center.</p>
<h4>Zoom lenses</h4>
<p>To be honest, this whole project started with a <a href="https://omar.website/posts/notes-from-dynamicland-geokit/zoom-lens-series.jpg">very clear
image</a> I had in my head. I wanted to make <em>zoom
lenses</em>, where you could point a lens at a map to zoom into it, then
point a lens at that lens... so I want to show you those lenses now.</p>
<div>
<p>Applying a zoom lens to the inset map, then
applying a zoom lens to that zoom lens</p>
</div>
<p>You can see me playing with the OpenPTMap (public transport) tilelayer
as I mess with zoom, applying it here and there at various zoom levels
(it propagates down from wherever I put it). The layer shows some bus
lines, as well as the J-Church light rail line.<sup></sup> I'm
trying to get the <em>flexibility</em> of these tools across to you: layers
work everywhere and zoom works everywhere, and you can pull a tool out
and use it wherever and whenever you want.</p>
<p>Here's my play on <a href="https://en.wikipedia.org/wiki/Powers_of_Ten_(film)">Powers of
Ten</a> in Geokit:
starting from the Earth, I apply a zoom lens, then a lens to that
lens, and so on, until I get down to our neighborhood in Oakland.</p>
<p>I can even print a map. I have a printed map of San Francisco here,
for instance. I use exactly the same tools on this printout: I look at
public transit lines, then demographics, then I zoom and pan around
east of Golden Gate Park.</p>
<div>
<p>Working with a printed map: applying tilelayers
and zoom lens
</p>
</div>
<p>Even if Geokit isn't drawing any tiles on a map and it's just a
printout, it's still a fully functioning map.</p>
<p>I've been using San Francisco as an example, but I could have entered
any place in the world, and everything would have worked exactly the
same.<sup></sup></p>
<p>The one-page 'engine' page makes all these maps and tools work, and
that little engine doesn't need to know anything about the stuff built
above it. Basically, the engine takes a page representing some
geographic coordinates, and it draws some map tiles from the Internet
on top. I'll show how you might build that engine from scratch here in
Dynamicland, step by step.</p>
<h3>All the parts</h3>
<p>These are all the 'game pieces' inside the kit. Each page is both a
little computer program and an <em>object</em> that people and other pages
can see and work with. Some pages are face-down, but most of them are
face-up, which means they're running right now. (Face-up is the side
with the colored dots.)</p>
<p>You often <strong>point</strong> one page at another to make it do something, like
how the <em>zoom lens</em> points at the <em>printed map</em>, or how the <em>keyboard</em>
points at the <em>text box</em> which points at the <em>big map</em>.</p>
<p>We have a few ways to get a map of a place:</p>
<ul>
<li>⌨️<strong>type a place name</strong> into the text box to search for it</li>
<li>use a premade map like my <strong>📄 printed map</strong> of SF</li>
<li>apply the <strong>🔍 zoom lens</strong> to an existing map: the lens becomes a zoomed-in map
of the green region it points at</li>
<li>use the <strong>🎛 zoom &amp; pan dial</strong> ('loupe') page on an existing map: it zooms into
wherever the dial is, and you rotate the dial to change the zoom
level</li>
</ul>
<p>These parts are the ones in the kit right now, but anyone here can
make new parts: you could come here and write up a new kind of
tilelayer or a new way to make maps.</p>
<p>That's why I call Geokit a kit and not an application: it's naturally
extensible. You could come here, pull a keyboard off the shelf, point
it at any part of Geokit, and just start editing it and printing new
stuff for your own use. You might provide a new tileset to answer some
question you have, or come up with a new zoom mechanism to derive
maps. (I'll explain how to do both of those things in this writeup!)
You have the same status that I do, and you don't have to open your
laptop and download and recompile some source code if you want to
change something.</p>
<h2>Background on Realtalk</h2>
<p><a href="https://dynamicland.org/">Dynamicland</a> is a place – a land – in
Oakland, California.</p>
<p><a href="https://harc.ycr.org/project/realtalk/">Realtalk</a> is the programming
system that runs inside Dynamicland. It includes a superset of the
<a href="https://en.wikipedia.org/wiki/Lua_(programming_language)">Lua programming
language</a>. You
do 'pure' computation in Lua, but you use special Realtalk extensions
to do all communication with other pages and with the outside world.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/realtalk-cheat-sheet.png"><p>An overview of Realtalk concepts and syntax, by
<a href="https://twitter.com/telogram">Tabitha Yong</a></p>
</div>
<p>The Realtalk 'operating system' already knows about some wishes as
part of its standard library, so we can say things like <code>Wish (you) is labelled "Bella".</code> and <code>Wish (you) is highlighted "green".</code> and they
work right away. But we can also make up our own claims and wishes:</p>
<pre tabindex="0"><code>-- Made-up claim demo
Claim (you) blahblahblah.

-- Made-up claim when demo
When /page/ blahblahblah:
    Wish (page) is highlighted "blue".
End
</code></pre>
<div>
<a href="https://omar.website/posts/notes-from-dynamicland-geokit/made-up-claim.jpg"><img src="https://omar.website/posts/notes-from-dynamicland-geokit/made-up-claim.jpg"></a>
<p>Acting on claim syntax that I just made up (<a href="https://omar.website/posts/notes-from-dynamicland-geokit/made-up-claim.jpg">full size</a>)</p>
</div>
<p>Why is "Made-up claim demo" highlighted in blue? What will happen if I
flip "Made-up claim when demo" face-down?</p>
<h3>Realtalk and the rest of Dynamicland</h3>
<p>I'm mainly going to discuss Realtalk here, so I'm writing in a form
that will look almost like any other tutorial about how to do a
programming project.</p>
<p>I'm not going to talk about the <a href="https://dynamicland.org/research-notes/social-dynamics-of-programming-together/">social experience of
Dynamicland</a>,
the community, or things particular to the physical lab, but those
things are integral to the Dynamicland concept! Knowledge transfer in
Dynamicland usually happens those ways, not through writing like
this. I just think others can do a better job of describing those
things, and I want to fill the gap of documenting 'programming.'
Check out <a href="https://twitter.com/dynamicland1">@Dynamicland1</a> on Twitter
for a broader view.</p>
<h2>Making Geokit</h2>
<p>I'll start by creating a map page representing San Francisco, a map
page which won't actually do anything at first.</p>
<p>Then I'll build up the Geokit 'engine', which is about a page of code,
in four steps. (The engine will make the map page work.)</p>
<ol>
<li>
<p>A starter three-line Geokit 'engine page' that just paints that map blue.</p>
</li>
<li>
<p>A better engine which downloads a map tile from OpenStreetMap and
displays it.</p>
</li>
<li>
<p>Support for multiple tile layers, from different sources, on one
map.</p>
</li>
<li>
<p>Stitching more than one tile together to get exactly the right map
view to fit the map page's area.</p>
</li>
</ol>
<h3>A map page that doesn't do anything (yet)</h3>
<p>First, I'll make a San Francisco map page for Geokit to act on.</p>
<p>I want to make a claim that my map page represents the geographical
area of San Francisco, so that Geokit knows it's supposed to draw San
Francisco on it. I'm going to type in a claim whose syntax I made up:
<code>Claim (you) is geomap of bbox (... coordinates ...).</code></p>
<p>At Dynamicland, you make a new page by</p>
<ol>
<li>pointing a keyboard at an existing page someone else wrote (ideally
with code you want to reuse),</li>
<li>then patching its code in place to do what you want,</li>
<li>and then printing it out.</li>
</ol>
<p>I'll do that to make my map page, although I'm not keeping any of the
original code in this case. I also add a bunch of linebreaks so the
printout covers the whole page.</p>
<p>
</p>
<p>Is this a map? Well, it's a page – a program, basically – which is
making a claim about itself. (See <a href="https://omar.website/posts/notes-from-dynamicland-geokit/realtalk-cheat-sheet.png">Tabitha Yong's cheat sheet</a>, shown
earlier, for more about claims.)</p>
<div>
<a href="https://omar.website/posts/notes-from-dynamicland-geokit/san-francisco-map-page.png"><img src="https://omar.website/posts/notes-from-dynamicland-geokit/san-francisco-map-page.png"></a>
<p>Our map page (<a href="https://omar.website/posts/notes-from-dynamicland-geokit/san-francisco-map-page.png">full size</a>). It's
the 17,595th page!</p>
</div>
<pre tabindex="0"><code>-- San Francisco map
Claim (you) is geomap of bbox ({left=-122.527, bottom=37.664, right=-122.348, top=37.851}).
</code></pre>
<p>You can think of it as setting a dynamic 'property' that other pages
can see: "I claim I'm a map representing the box with corners
(122.527° W, 37.664° N) and (122.348° W, 37.851° N)." (Those
longitude-latitude pairs are, very roughly, the bounds of San
Francisco.)</p>
<div>
<a href="https://omar.website/posts/notes-from-dynamicland-geokit/show-claims-and-wishes.jpg"><img src="https://omar.website/posts/notes-from-dynamicland-geokit/show-claims-and-wishes.jpg"></a>
<p>A debug aid: showing claims made by each page on
the table (<a href="https://omar.website/posts/notes-from-dynamicland-geokit/show-claims-and-wishes.jpg">full size</a>)</p>
</div>
<p>You can make whatever claims you want, but there's no guarantee that
anything will <em>happen</em> as a result. Right now, nobody cares about my
geomap claim, so the map is inert. But soon, the Geokit 'engine' will
be querying for all the pages that make that kind of claim, and it
will give them map behavior.</p>
<h3>1. Starting on the 'engine' page</h3>
<p>Next, I'll start building the Geokit 'engine', which is a separate
page: it'll look around for maps like the one I just made, then draw
on them.</p>
<p>My first attempt is a minimal test to show how I can recognize
and act on maps: it colors all maps solid blue.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/geokit-1st-try-blue.png"><p>Geokit engine page: colors all maps blue</p>
</div>
<pre tabindex="0"><code>-- Geokit 1st try: blue
-- On each geomap, draws blue.

When /map/ is geomap of box /bbox/:

    Wish (map) is highlighted "blue".

End
</code></pre>
<p>I'm using a wish here, which is a sort of broadcast request for
something in the world to be different: in this case, whenever I see a
map, I wish for it to be highlighted in blue.</p>
<p>Some other page happens to notice that wish and fulfill it using a
projector, but <a href="https://twitter.com/worrydream/status/1024496333200056320">I don't know or care about that
mechanism.</a>
I'm not addressing the wish to anyone in particular. When I put this
page together with the map, sure enough, the map turns blue.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/apply-geokit-1st-try-blue.jpg"><p>Turning the engine on and coloring SF blue</p>
</div>
<p>To show you that the engine responds to every map, not just one, I
made a map of New York as well. See for yourself: it turns blue the
same way.</p>
<div>
<video></video><p>Making both New York and SF blue</p>
</div>
<h3>2. Drawing one map tile</h3>
<p>OK. Now I want to draw <a href="https://wiki.openstreetmap.org/wiki/Tiles">actual map tiles from
OpenStreetMap</a> instead of
that placeholder blue. I'll start with something small: draw <a href="https://tile.openstreetmap.org/11/327/791.png">one
tile of SF</a> at a fixed
level of detail, big enough that it fits the page.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/apply-geokit-2nd-try.jpg"><p>Drawing a map of SF!</p>
</div>
<p>I'll show you (an abbreviated version of) how I edited our previous
"paint maps blue" 1st try and ended up with this 2nd try, which draws
an actual map tile. I'll show you the code next, so don't worry about
squinting to read all the code in the video. (The projection is much
clearer in person!)</p>
<p>I start by changing <code>"blue"</code> to <code>"orange"</code> and then <code>"red"</code> to show
you how I can edit the page in place. Then I extend the code so that
it downloads and draws a hard-coded tile at <code>(11, 327, 791)</code>, which is
San Francisco. Finally, I make the code more general. I copy and paste
the utility function <code>deg2num</code> from another page, which I use to
compute a tile based on the <code>bbox</code> encoded in the map page, along with
a drawing size for that tile based on the page's physical height and
width. Once it works, I print the page out, so I have a permanent,
physical version.</p>
<div>
<p>Editing Geokit 1st try into Geokit 2nd try, which
actually renders a map. Printing the final 2nd try out at the end</p>
</div>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/geokit-2nd-try-one-tile.png"><p>Geokit engine page, 2nd try: draws one OSM tile on
each map. Key When block is highlighted</p>
</div>
<p>This code is some of the most 'traditional' and complicated in the
whole kit. Honestly, it's a little unidiomatic to have so much
'normal' code bunched together. It doesn't help that it has to
interface with the outside world (the Internet and filesystem).</p>
<p>You might start by looking at the code block I highlighted, the <code>When /map/ is geomap...</code> block at the bottom, which runs for each map on
the table:</p>
<pre tabindex="0"><code>local URL_TEMPLATE = "https://tile.openstreetmap.org/%d/%d/%d.png"
local ZOOM = 11
When /map/ is geomap of bbox /bbox/,
     /map/ has width /width/, /map/ has height /height/:

    local xtile, ytile = deg2num((bbox.left + bbox.right)/2,
                                 (bbox.top + bbox.bottom)/2,
                                 ZOOM)
    local url, path = get_tile_url_and_path(URL_TEMPLATE, ZOOM,
                                            math.floor(xtile),
                                            math.floor(ytile))
    local tilesize = math.min(width, height)
    When tile URL (url) is downloaded to path (path):
        local ill = new_illumination()
        ill:image { filename=path, scale=1,
                    x=0, y=0,
                    width=tilesize, height=tilesize }
        Wish (map) has illumination (ill).
    End

End
</code></pre>
<ol>
<li>
<p>Have the <code>When</code> bind variables:</p>
<ul>
<li>
<code>map</code>: reference to the map page</li>
<li>
<code>bbox</code>: the left, right, top, bottom coordinates of this geomap</li>
<li>
<code>width</code>: the map page's usable width in inches (<code>7.5</code>)</li>
<li>
<code>height</code>: the map page's usable height in inches (<code>10</code>)</li>
</ul>
</li>
<li>
<p>Compute the geographic center of <code>bbox</code> and use
<a href="https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Lua"><code>deg2num</code></a>
to find the corresponding OSM tile coordinates <code>xtile</code> and <code>ytile</code>
(<code>327</code> and <code>791</code>, respectively, in this case)</p>
</li>
<li>
<p>Construct the URL <a href="https://tile.openstreetmap.org/11/327/791.png">https://tile.openstreetmap.org/11/327/791.png</a> and
the download path
<code>shared/geokit-tiles/https___tile_openstreetmap_org_11_327_791_png.png</code></p>
</li>
<li>
<p>Compute the <code>tilesize</code> at which to render the tile image, which is
the minimum of the page width and height. In this case, it's about
<code>7.5</code> inches.</p>
</li>
<li>
<p>Once the tile is downloaded<sup></sup>, build an <em>illumination</em> (the
built-in graphics library) that draws the tile image. Then <code>Wish</code> to
draw that illumination on the map page.</p>
</li>
</ol>
<p>I patched in some log statements to show you the live values of
<code>bbox</code>, <code>width = 7.491</code>, <code>height = 10.001</code>, <code>xtile = 327</code>, and <code>ytile = 791</code>:</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/geokit-2nd-try-logs.jpg"><p>Viewing live values with log statements</p>
</div>
<h3>3. Modular tile layers</h3>
<p>I can make a ten-line change to the engine and get a huge amount more
flexibility. Instead of hard-coding that <code>URL_TEMPLATE = "https://tile.openstreetmap.org/%d/%d/%d.png"</code> in Geokit, I'll let
pages make <code>Wish</code>es to draw any tilesets from any URLs they want.</p>
<p>To let the user specify what layers to draw, I've made individual
tilelayer pages. Each page represents a layer: when you point a layer
page at a map page, the layer wishes to draw that layer on the map.</p>
<p>Here's the source code for the "Demographic dots tilelayer", for
example:</p>
<pre tabindex="0"><code>When (you) points "up" at /map/:
    Wish (map) shows tilelayer
        "http://demographics.virginia.edu/DotMap/tiles4/%d/%d/%d.png"
        with options ({max_zoom=13}).
End
</code></pre>
<p>I edited the engine so that it responds to these <code>Wish</code>es, rather than
looking for any <code>geomap</code> and drawing that hard-coded <code>URL_TEMPLATE</code>
from last time.</p>
<p>Here are all four tilelayers in action on the San Francisco
map. Notice how I chain layers to put them on top of each other!</p>
<p>And here is the new engine page:</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/geokit-3rd-try-tilelayers.png"><p>Geokit engine page: allows custom tilelayers for
each map. Changes from 2nd try are highlighted</p>
</div>
<p>This 3rd-try Geokit is almost unchanged from the 2nd try, except for
this section, from line 34 down to the extra clause in the key <code>When</code>
block that handles maps:</p>
<pre tabindex="0"><code>When /someone/ wishes /page/ shows tilelayer /url_template/ with options /opts/,
     /page/ wishes /target/ shows tilelayer /something/ with options /something/,
     page ~= target:
    opts = rt.table(opts) -- Make a mutable copy of opts
    opts.priority = 1 + (opts.priority or 1)
    Wish (target) shows tilelayer (url_template) with options (opts).
End

local ZOOM = 11
When /someone/ wishes /map/ shows tilelayer /url_template/ with options /opts/,
     /map/ is geomap of bbox /bbox/,
     /map/ has width /width/, /map/ has height /height/:
</code></pre>
<p>Start from the bottom: the added <code>When /someone/ wishes /map/ shows tilelayer /url_template/ with options /opts/</code> clause makes my
map-drawing code run separately, for every tilelayer wish for each
map, rather than running only once for each map.</p>
<p>It binds <code>url_template</code> to whatever the tilelayer page said, whereas I
had a hard-coded <code>URL_TEMPLATE</code> constant last time. The <a href="https://omar.website/posts/notes-from-dynamicland-geokit/osm-tilelayer.png">tilelayer
page</a> code makes a wish like <code>Wish (map) shows tilelayer "https://tile.openstreetmap.org/%d/%d/%d.png" with options ({})."</code>, which results in <code>url_template = "https://tile.openstreetmap.org/%d/%d/%d.png"</code> and <code>opts = {}</code>.</p>
<p>The <em>first</em> <code>When</code> block (<code>When /someone/ wishes /page/ shows tilelayer...</code>), starting on line 34, is really interesting. It's what
lets me put layers on top of each other by chaining them. When a
tilelayer page (<code>someone</code>) is pointed at another tilelayer page
(<code>page</code>), and that <code>page</code> is itself pointed at <code>target</code>, then
<code>someone</code>'s tilelayer wish is 'forwarded' to <code>target</code>, but with its
priority boosted by 1 (higher priority means drawn on top, so
<code>someone</code>'s tilelayer gets drawn above <code>page</code>'s tilelayer). That means
that a tilelayer pointed directly at the map is drawn at the bottom, a
tilelayer pointed at <em>that</em> tilelayer is in front of it, and so
on. The outermost tilelayer is drawn on top.</p>
<h4>A note on the Realtalk model</h4>
<p>Suppose you were implementing this tilelayer manipulation in
traditional software. You'd probably make a list of tilelayers, with
(+) and (-) buttons. You might put a check box next to each
tilelayer. You'd have to implement drag and drop, and you'd have to
handle events every time the user reordered the
layers.<sup></sup></p>
<p>This step starts to illustrate the power of the Realtalk model, in a
way that <a href="https://twitter.com/rsnous/status/1018570015866384389">I've tweeted about
before</a>. Look
at how I've handled layers. Each layer is a page, and you put the
layer down and point it to enable it and set its position. Nobody had
to implement a checkbox or handle a drag event or synchronize an array
with a UI view.</p>
<p>When I first showed up at Dynamicland, I did a lot of what I think of
as 'traditional programming.' That means using a page essentially as a
<em>computer monitor</em>: you hack out a lot of Lua code, print it, and it
renders something cool on that page. Maybe it takes input based on
different regions of the page. I've seen programmers do this over and
over on their first exposure to the system, and I fell into it too: I
would say it took me about four months to understand how to really
take advantage of Realtalk.</p>
<p>One way to think about Realtalk is that a lot of code in the outside
world is about <em>displaying lists</em> and <em>letting people pick from
lists</em>.</p>
<ul>
<li>
<p>Mapping software has to show a list of layers, and make them
checkable and uncheckable, and make them draggable, and handle all the
events, and remember what the state is.</p>
</li>
<li>
<p>A photo app shows you a list of photos, and then you open one, and
then you see a toolbox of operations you can apply, and then you pick
one...</p>
</li>
<li>
<p>Even command-line programs let you ask for a list of source branches
or a list of disk partitions or a list of files or whatever, and have
some set of subcommands you can apply to your object of choice.</p>
</li>
</ul>
<p>In Realtalk, most of that code falls away – provided that you
represent your objects as individual physical pages. Once you have
physical pages, <strong>you automatically get the operations of the physical
world</strong>: placing and picking up objects, moving and grouping objects
in space, pointing objects at each other, and so on.</p>
<p>We have a rule of thumb that no program should be longer than a page
(73 lines!) If you have a longer program, you're probably doing
something wrong, and you should split it up into little pages that
work together.</p>
<p>And there's a lighter rule of thumb floating around, that most code
should be 'purple text' – <code>When</code>s and <code>Claim</code>s and <code>Wish</code>es, which the
editor highlights in purple – not gobs of plain Lua code.</p>
<p>These rules suggest that there's a distinct Realtalk idiom, which is
very different from just writing Lua for projectors and cameras. It's
something like objects or actors (where the pages are objects), and
reactive programming, and reusing functionality from the real world
wherever possible.</p>
<p>I don't claim that Geokit is an ideal example: the engine page is
pretty complex, and even my use of the projector to render maps is
questionable, since any static map could have been printed
instead. But the layer pages show a little bit of this power.</p>
<h3>4. Fitting to exact coordinates</h3>
<p>Finally, I want to get rid of that one-tile hack. If a map page is
supposed to represent the box with corners (122.527° W, 37.664° N) and
(122.348° W, 37.851° N), I want it to display a map of <em>that exact
area</em>, not just some nearby tile image that OpenStreetMap serves
up.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/before-perfect-fit.jpg"><img src="https://omar.website/posts/notes-from-dynamicland-geokit/after-perfect-fit.jpg"><p>Comparing our previous engine (left) to the
behavior I want (right). Hover over right to see a patched view with tile boundaries
and coordinates.</p>
</div>
<p>We need to figure out the right level of detail, get multiple tiles at
that level, and then stitch the map together.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/geokit-4th-try-perfect-fit.png"><p>Geokit engine page: draws tiles to fit the map's
bounding box. Changes from 3rd try are highlighted</p>
</div>
<p>I've made this engine development into a <a href="https://gist.github.com/osnr/773a04aa8ce405a87d2122a9547b73df">GitHub
gist</a>
as well, so you can look at the
<a href="https://gist.github.com/osnr/773a04aa8ce405a87d2122a9547b73df/revisions">revisions</a>
to see how I got from step 1 to here.</p>
<p>I've only changed the bottom <code>When</code> block from the previous engine:</p>
<pre tabindex="0"><code>
When /someone/ wishes /map/ shows tilelayer /url_template/ with options /opts/,
     /map/ is geomap of bbox /bbox/,
     /map/ has width /width/, /map/ has height /height/:

    -- Made-up formula to guess a good zoom from width of bbox.
    local zoom = math.log(360 / (bbox.right - bbox.left)) / math.log(2)
    zoom = math.ceil(zoom + width/16 - 1) -- experimentally determined width factor
    local max_zoom = opts.max_zoom or 17
    if zoom &gt; max_zoom then zoom = max_zoom end
  
    local xtile_left, ytile_top = deg2num(bbox.left, bbox.top, zoom)
    local xtile_right, ytile_bottom = deg2num(bbox.right, bbox.bottom, zoom)

    local tilesize = width / (xtile_right - xtile_left)

    for xtile=math.floor(xtile_left),math.floor(xtile_right) do
        for ytile=math.floor(ytile_top),math.floor(ytile_bottom) do
            local url, path = get_tile_url_and_path(url_template, zoom, xtile, ytile)
            When tile URL (url) is downloaded to path (path):
                local ill = new_illumination()
                ill:image { filename=path, scale=1,
                            x=tilesize * (xtile - xtile_left),
                            y=tilesize * (ytile - ytile_top),
                            width=tilesize, height=tilesize }
                Wish (map) has illumination (ill) with priority (opts.priority or 1).
            End
        end
    end
End
</code></pre>
<ol>
<li>
<p>Determine the <a href="https://twitter.com/rsnous/status/1017898981349732353">zoom
level</a>
<code>zoom</code> – higher zoom means <a href="https://wiki.openstreetmap.org/wiki/Zoom_levels">more tiles are used to represent the
same area</a>. I
should point out that there's no single <em>correct</em> zoom level to
draw a particular map: it depends on how much detail you want to
show. You probably want higher zoom the higher your display
resolution is.</p>
<p>Here, I take the logarithm of (360 / the distance between the left
longitude and right longitude of the map), so a map covering a
smaller geographical area gets higher zoom. Then I increase the
zoom by a bit, based on the physical size of the map page <code>(width / 16)</code>. For this San Francisco map, I end up with <code>zoom = 11</code>.</p>
</li>
<li>
<p>Compute the <code>tilesize</code> in inches at which to render each tile, such
that the tiles exactly fit the page.<sup></sup> In this case, we get something
like <code>tilesize = 7.44</code> inches.</p>
</li>
<li>
<p>Loop through all the tiles and download and render each one, just
as we did when we drew a single tile previously.</p>
</li>
</ol>
<p>With perfect fit working, maps of different sizes also work
great. Here's the New York map I made earlier.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/with-new-york.jpg"><p>New York and SF. Hover to see tile boundaries</p>
</div>
<p>I showed a lot of tools, like zooming and searching for places by
name, at the beginning; how do they work?</p>
<p>You can implement all of them without changing the engine at all. I
won't go into as much detail about them – maybe some other time – but
they all fundamentally come down to the claim that a page <code>is geomap of bbox (...something...)</code>. I've used constant coordinates for the
<code>bbox</code> so far, but they can also be dynamic values, which is how we
get dynamic maps.</p>
<h3>Zoom lens</h3>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/zoom-lens-series.jpg"><p>Applying 2 zoom lenses to get from the whole US to
Northern California</p>
</div>
<p>The zoom lens is surprisingly simple to implement, and it was the
first tool I made. Wanting to make a zoom lens inspired the whole
project, after all.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/zoom-lens.png"><p>Zoom lens page: zooms into map to its left</p>
</div>
<p>The lens draws the green rectangle on the map it's zooming into, then
computes a <code>bbox</code> so it can make its own claim to be a geomap.</p>
<p>It inherits tilelayers from its parent, but scales their priorities
down so the user can easily put layers on top. It also inherits
highlighted points, which are an addon I'll discuss next.</p>
<h3>Addons</h3>
<p>I need some addons to support the rest of the tools.</p>
<h4>Wish (...) is geomap containing bbox (...)</h4>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/addon-wish.png"><p>Addon to handle geomap wish</p>
</div>
<p>This addon handles <code>Wish (...) is geomap containing bbox (...).</code> It
lets a page become a map of an area that doesn't exactly match its
aspect ratio. I'll need this functionality soon, when I want to map an
arbitrary place that the user types in.</p>
<p>For example, if you have a <code>page</code> with a 2:1 aspect ratio, but you
want to make <code>page</code> a map of square San Francisco, you can do <code>Wish (page) is geomap containing bbox (...SF coordinates...)</code>, and this
addon will give <code>page</code> a claim for a wider bbox which still contains
San Francisco.</p>
<p>If you already knew a <code>bbox</code> that matched the <code>page</code>'s 2:1 aspect
ratio, then you wouldn't need this wish addon: you could simply make
the original <code>Claim (page) is geomap of bbox (...).</code> that we
implemented in the engine.</p>
<h4>Wish (...) highlights geopoint (...) with label (...)</h4>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/addon-highlight.png"><p>Addon to handle highlighting points</p>
</div>
<p>This addon handles <code>Wish (...) highlights geopoint (...) with label (...).</code> You can label specific places on a map with this
<code>Wish</code>.</p>
<h3>Place search</h3>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/place-search-example.jpg"><p>Typing in "USA" and getting a map</p>
</div>
<p>Place search goes to Google Maps with whatever you typed, gets a
suggested viewport back, and then makes a <code>Wish</code> (to the geomap-wish
addon from earlier) to construct a map containing that viewport. It
also sets the OSM tilelayer as a default.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/place-search.png"><p>"Geomap above from place search": point keyboard
at it and type, and this page turns the page above it into a map</p>
</div>
<h3>Zoom and pan dial</h3>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/ir-dial-example.jpg"><p>Zooming into part of Queens with the "IR
dial." Notice the 4 retroreflective dots on the dial, which reflect
infrared light back to a camera, so I can track the dial's position and angle</p>
</div>
<p>This dial pan-and-zoom mechanic was suggested and implemented by some
visiting designers from a technology company you've heard of. We
showed off a few things in the morning, including a <a href="https://twitter.com/acwervo/status/983107535312797697">generic
dial</a> and my
Geokit zoom lenses, and one designer (who "didn't program")
immediately thought about how to combine the two.</p>
<p>How could we control the zoomed-in map <em>directly</em> and in context,
instead of having the lens (which sits at a distance from the original
map)? With that interaction idea and a couple of hours of work in the
afternoon, we had a working prototype for this dial. And we were
working at a table together, with pieces of paper, a marker for
scribbling, a dial, and a keyboard, not with laptops.</p>
<p>I love the direct manipulation; just as hoped, moving the dial on the
map to pan feels much nicer than moving the zoom lens out on the side.</p>
<p>The green number on the dial (from 0.0 to 1.0) is the zoom level,
where 1.0 is the most zoomed-in and 0.0 is the least.</p>
<p>There's also a green rectangle, which represents the exact area that
the inset map is showing. When I increase the zoom level, the green
rectangle shrinks and the inset map zooms in.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/ir-dial.png"><p>"Map magnifier with IR dial": zooms into map to its
left. Expects IR dial to be on that page; turn the dial to zoom and
move it to pan</p>
</div>
<p>The original dial was just a (small) normal page, which you'd put on
top of the map page. But the light from the projector interfered with
dot detection, especially with the racial dot tilelayer's bright
colors in place, so it was unreliable. To avoid that problem, I'm
using an IR (infrared) dot detector (<code>/map/ contains tracking dots /dots/</code>) which tracks the gray-looking retroreflective dots. (Ideally,
the dial would be a normal page, though.)</p>
<div>
<video></video><p>Looking at the underside of the IR dial
peripheral. How does it spin?</p>
</div>
<p>One of the best things about Dynamicland is making peripherals and
using them with the computer right away. We don't need to implement
simulated inertia or make a virtual drag handle, since we simply track
the physical dial.</p>
<p>The dot furthest from the average of the dots is the 'head' of the
dial, which I use to figure out the dial's angle.</p>
<p>It'd also be interesting to draw the zoomed map on the dial itself;
we've played with this a few times, but there are still some issues
(interference, performance, rotating view properly).</p>
<h3>Printing maps</h3>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/printed-map.jpg"><p>A printed map of SF. Notice the claim at the top</p>
</div>
<p>Printing is very much a <strong>work in progress</strong>. You point the "Print
static map" page at a map page, and it prints a static map of that
place, with a Google Maps view on it.</p>
<div>
<img src="https://omar.website/posts/notes-from-dynamicland-geokit/print-static-map.png"><p>"Print static map"</p>
</div>
<p>There's a lot of fiddly calibration to do here, and it's not perfect
right now. It goes to Google's Maps API and gets an image around the
center of the map it's pointed at, guessing the right radii around
that image to line up with the existing map. It injects a claim, so
the printed map will be a valid Geokit map too.</p>
<p>It's been surprisingly useful as a debugging tool, since I can turn a
misbehaving dynamic map, which is hard to keep stable, into a static,
printed test case.</p>
<h2>Conclusion</h2>
<p>These pages are all there is! Everything I showed you at the beginning
of this writeup was implemented by the actual pieces of paper in front
of me, and now you've seen how all that code works.</p>
<p>Yeah, I'm relying on some 'built-in' functionality, but all that
behavior is <em>also</em> implemented on pages here in Dynamicland. When
you're here, you can walk over to a board and read through those
system libraries. There is no underlying stack hidden on a computer
hard drive or on GitHub. On this whiteboard, you can find the page
that renders to the projector when I <code>Wish (map) has illumination...</code>,
the page that makes a GET request when I query <code>When URL (...) has contents /contents/:</code>, and the page that parses those <code>Claim</code>s and
<code>Wish</code>es and <code>When</code>s in the first place:</p>
<p>In Dynamicland, software is supposed to be this simple or
simpler. Software in Dynamicland is meant to be continually read and
tweaked and discarded by users; that's why each page has its source
code printed on it.</p>
<h3>Why maps?</h3>
<p>Why maps? Personally, I've been interested in maps and cities and
public transportation and infrastructure for a few years, and I
haven't gotten to do much original work around that interest, so
Dynamicland seemed like a good place to explore it. I was excited to
build an idiomatic-feeling 'kit' with parts, instead of a one-off
program like my earliest work here.</p>
<p>I think maps highlight the advantages of breaking out of the
screen. We insult an entire tradition of large, information-rich maps
when we cram them onto tiny phone screens, then crowd around them and
<a href="http://worrydream.com/ABriefRantOnTheFutureOfInteractionDesign/">make finger
gestures</a>
to squeeze answers out bit by bit.</p>
<p>I figured that in Dynamicland, we could have maps that actually looked
like maps. We could have multiple people sitting around a map at a
table, sharing a large view, but also applying computational filters
and derivations to get at their individual interests. You could
explore many places and ideas at once, and keep track of them on the
table in front of you, instead of constantly sacrificing context on a
limited screen.</p>
<p>Dynamicland is great at spatial interfaces, and maps are
spatial. You get tools like dot tracking, page tracking, whiskers, and
rotation, as well as whatever else you can tie in from the physical
world.</p>
<p>And maps involve real information from the world, which makes them
more relatable than a visualization of some mathematical model. I've
seen plenty of people whip out lenses and turn zoom knobs and poke at
layers in an attempt to learn about the place where they grew up. Maps
are an early opportunity to use this technology to ask real questions
about the world, the same way that I use my phone or laptop to ask
questions.</p>
<h3>On documentation</h3>
<p>I think a lot of the material and discussion about Dynamicland has
been either too vague or too focused on the wrong details. It's been
hard to talk about the place to people who haven't visited. You lapse
into phrases that sound like platitudes.</p>
<p>I'm taking a different approach here. I want to be concrete about how
you make things and about the details of the programming
experience. I'm hoping that this detail will actually do a better job
of communicating the bigger ideas.</p>
<p>I think there are a few reasons why there hasn't been much written
about Realtalk, especially writing beyond 'first impressions':</p>
<ul>
<li>
<p>Written essays feel like an awkward medium for talking about
Dynamicland. Because it's a physical space, its inhabitants are used
to communicating in person, with demonstrations at hand. You
wouldn't ordinarily write or read a document like this one; if you
want to know how something works, you put it out on the table and
read the code on the pages, and maybe ask someone in the space
(since it is a shared space, people will be there!)</p>
</li>
<li>
<p>Building external attention hasn't been a big goal: the focus has
been on in-person outreach, fundraising, and development within the
space.</p>
</li>
<li>
<p>Even among people who've visited Dynamicland, not many actually know
Realtalk well, so not many can write this kind of overview. As with
any creative tool, becoming fluent takes time. It took me at least a
few months.</p>
</li>
<li>
<p>There may also be some embarrassment about having a traditional
text-editor-oriented environment, given the earlier work of
<a href="http://worrydream.com">Bret</a> and other researchers in the lab.</p>
</li>
<li>
<p>There was a fear that people would fixate on details and copy them
blindly while missing the values, as happened to the Xerox Alto GUI.</p>
</li>
</ul>
<p>But focusing on these programming details still seems better than
focusing on things like the dots and projectors and computer vision,
which are obvious to anyone who visits and so get a ton of
discussion. I see that stuff – dots, the nature of the page database,
cameras and projectors, occlusion, comparisons to augmented reality –
purely as implementation details, and I don't spend much time thinking
about it; I haven't discussed it in this writeup at all!</p>
<p>The pages/wishes/claims/whens programming model is the true 'spirit of
Realtalk', even if ideally we'd change programming practice in even
more radical ways: get rid of the keyboards as well, or add more
direct manipulation, or introduce better visibility for
debugging... or abolish programming entirely?</p>
<div>
<a href="https://omar.website/posts/notes-from-dynamicland-geokit/nicky-case-programming.jpg"><img src="https://omar.website/posts/notes-from-dynamicland-geokit/nicky-case-programming.jpg"></a>
<p>One suggestion for what 'programming' in
Dynamicland could look like,
by <a href="https://ncase.me">Nicky Case</a> (<a href="https://omar.website/posts/notes-from-dynamicland-geokit/nicky-case-programming.jpg">full size</a>)</p>
</div>
<p>In fact, I'd argue that the programming model is very
under-discussed. It is more important than you may think from what you
see online. A lot of what's new about Dynamicland is that <em>it's a
programming system!</em> Compare projection-mapping demos you see from the
1990s and 2000s, from various academic and industrial labs. They may
have similar interface techniques, and they may demo cleaner, but the
implementation is some big blob of C++ that is written backstage. You
can't walk in and do your own program. The pieces of paper in
Dynamicland are not just display surfaces for a demo someone wrote on
their PC: they're used to program the system directly.</p>
<p>You could implement Geokit in JavaScript or C++, and build your own
projection mapping and computer vision rig, and you could implement
<em>any</em> Dynamicland program that way, but it would be a lot of work. I
certainly would never have done it. Dynamicland lets you take all that
as given.</p>
<p>There's also lots of Dynamicland coverage that is just out of date!
Using the system today (August, 2018) is a very different experience
compared to using the system a year or even six months ago: there are
thousands of example programs, computing surfaces are everywhere in
the space, projectors are high-resolution enough to display detailed
data and text, everything runs way faster, you edit with keyboards
within the system instead of laptops – the fundamental design is
intact, but the subjective experience has changed a lot, especially
when you're making things.</p>
<p>I hope this writeup has given you a better idea of how Dynamicland
works and what's powerful and new about it. I've talked about some of
these ideas <a href="https://twitter.com/rsnous">@rsnous on Twitter</a>, and I'd
like to write and release more in public – let me know what you want
to hear about.</p>
<p>best,<br>
Omar</p>
<p><em>My thanks to John Backus, Jason Brennan, Nicky Case, Mary Rose Cook,
Daniel Feichtinger, Katie Gu, Jacob Haip, Joshua Horowitz, Luke
Iannini, Lauren Killingsworth, Tristan Navarro, Lisa Neigut, Paul
Sonnentag, Bret Victor, Roshan Vid, Elizabeth Yang, and Andy Zhao for
their thoughtful comments and feedback on this writeup.</em></p>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Omar Rizwan <br>
<br>
January 16, 2024 at 09:52AM
